[bug] [typescript] Projects using @klasa/core do not compile

Due to the main code having some `// @ts-ignore` comments which are not carried over to the build `.d.ts` files, projects currently using TypeScript and `@klasa/core` do not compile. Trying to compile a project that is basically the ping example from the readme (sample repo: https://bitbucket.org/Favna/klasa-core-test/) gives the following errors:

```
$ tsc -b src
node_modules/@klasa/core/dist/src/lib/pieces/base/Store.d.ts:71:5 - error TS2416: Property 'set' in type 'Store<V>' is not assignable to the same property in base type 'Cache<string, V>'.
  Type '(piece: V) => V | null' is not assignable to type '(key: string, value: V) => this'.
    Types of parameters 'piece' and 'key' are incompatible.
      Type 'string' is not assignable to type 'V'.
        'V' could be instantiated with an arbitrary type which could be unrelated to 'string'.

71     set(piece: V): V | null;
       ~~~

node_modules/@klasa/core/dist/src/lib/caching/stores/GuildEmojiStore.d.ts:57:15 - error TS2416: Property '_add' in type 'GuildEmojiStore' is not assignable to the same property in base type 'DataStore<GuildEmoji>'.
  Type '(data: APIEmojiData) => GuildEmoji' is not assignable to type '(data: { [k: string]: any; id: string; }) => GuildEmoji'.
    Types of parameters 'data' and 'data' are incompatible.
      Property 'name' is missing in type '{ [k: string]: any; id: string; }' but required in type 'APIEmojiData'.

57     protected _add(data: APIEmojiData): GuildEmoji;
                 ~~~~

  node_modules/@klasa/dapi-types/dist/index.d.ts:231:5
    231     name: string | null;
            ~~~~
    'name' is declared here.

node_modules/@klasa/core/dist/src/lib/caching/stores/GuildInviteStore.d.ts:43:15 - error TS2416: Property '_add' in type 'GuildInviteStore' is not assignable to the same property in base type 'DataStore<Invite>'.
  Type '(data: APIInviteData) => Invite' is not assignable to type '(data: { [k: string]: any; id: string; }) => Invite'.
    Types of parameters 'data' and 'data' are incompatible.
      Type '{ [k: string]: any; id: string; }' is missing the following properties from type 'APIInviteData': code, channel

43     protected _add(data: APIInviteData): Invite;
                 ~~~~

node_modules/@klasa/core/dist/src/lib/caching/stores/GuildMemberStore.d.ts:69:15 - error TS2416: Property '_add' in type 'GuildMemberStore' is not assignable to the same property in base type 'DataStore<GuildMember>'.
  Type '(data: MemberData) => GuildMember' is not assignable to type '(data: { [k: string]: any; id: string; }) => GuildMember'.
    Types of parameters 'data' and 'data' are incompatible.
      Type '{ [k: string]: any; id: string; }' is not assignable to type 'MemberData'.
        Type '{ [k: string]: any; id: string; }' is missing the following properties from type 'APIGuildMemberData': nick, roles, joined_at, deaf, mute

69     protected _add(data: MemberData): GuildMember;
                 ~~~~

node_modules/@klasa/core/dist/src/lib/caching/stores/PresenceStore.d.ts:27:15 - error TS2416: Property '_add' in type 'PresenceStore' is not assignable to the same property in base type 'DataStore<Presence>'.
  Type '(data: APIPresenceUpdateData) => Presence' is not assignable to type '(data: { [k: string]: any; id: string; }) => Presence'.
    Types of parameters 'data' and 'data' are incompatible.
      Property 'user' is missing in type '{ [k: string]: any; id: string; }' but required in type 'APIPresenceUpdateData'.

27     protected _add(data: APIPresenceUpdateData): Presence;
                 ~~~~

  node_modules/@klasa/dapi-types/dist/index.d.ts:490:5
    490     user: {
            ~~~~
    'user' is declared here.

node_modules/@klasa/core/dist/src/lib/caching/stores/VoiceStateStore.d.ts:27:15 - error TS2416: Property '_add' in type 'VoiceStateStore' is not assignable to the same property in base type 'DataStore<VoiceState>'.
  Type '(data: APIVoiceStatePartial) => VoiceState' is not assignable to type '(data: { [k: string]: any; id: string; }) => VoiceState'.
    Types of parameters 'data' and 'data' are incompatible.
      Type '{ [k: string]: any; id: string; }' is missing the following properties from type 'APIVoiceStatePartial': channel_id, user_id, session_id, deaf, and 4 more.

27     protected _add(data: APIVoiceStatePartial): VoiceState;
                 ~~~~

node_modules/@klasa/core/dist/src/lib/caching/stores/InviteStore.d.ts:37:15 - error TS2416: Property '_add' in type 'InviteStore' is not assignable to the same property in base type 'DataStore<Invite>'.
  Type '(data: APIInviteData) => Invite' is not assignable to type '(data: { [k: string]: any; id: string; }) => Invite'.
    Types of parameters 'data' and 'data' are incompatible.
      Type '{ [k: string]: any; id: string; }' is not assignable to type 'APIInviteData'.

37     protected _add(data: APIInviteData): Invite;
                 ~~~~

node_modules/@klasa/core/dist/src/lib/caching/stores/BanStore.d.ts:83:15 - error TS2416: Property '_add' in type 'BanStore' is not assignable to the same property in base type 'DataStore<Ban>'.
  Type '(data: { guild_id: string; user: APIUserData; }) => Ban' is not assignable to type '(data: { [k: string]: any; id: string; }) => Ban'.
    Types of parameters 'data' and 'data' are incompatible.
      Type '{ [k: string]: any; id: string; }' is missing the following properties from type '{ guild_id: string; user: APIUserData; }': guild_id, user

83     protected _add(data: GuildBanAddDispatch['d']): Ban;
                 ~~~~


Found 8 errors.

error Command failed with exit code 1.
```

Taking a closer look at the top one:
```
node_modules/@klasa/core/dist/src/lib/pieces/base/Store.d.ts:71:5 - error TS2416: Property 'set' in type 'Store<V>' is not assignable to the same property in base type 'Cache<string, V>'.
  Type '(piece: V) => V | null' is not assignable to type '(key: string, value: V) => this'.
    Types of parameters 'piece' and 'key' are incompatible.
      Type 'string' is not assignable to type 'V'.
        'V' could be instantiated with an arbitrary type which could be unrelated to 'string'.

71     set(piece: V): V | null;
       ~~~
```

This code in the source is, with a `// @ts-ignore` comment.
https://github.com/dirigeants/core/blob/a82ce4160b982df9d7c3d33f30ef49ae30afe357/src/lib/pieces/base/Store.ts#L117-L128

However the code that's in `dist/src/lib/pieces/base/Store.d.ts` does not (full file: https://unpkg.com/browse/@klasa/core@0.0.2/dist/src/lib/pieces/base/Store.d.ts):
```

/**
--
67 | * Sets up a piece in our store.
68 | * @since 0.0.1
69 | * @param piece The piece we are setting up
70 | */
71 | set(piece: V): V \| null;
```

That said, currently the only way for TypeScript users to use @klasa/core is by setting `skipLibCheck` to `true` in their tsconfig's. This is far from ideal and it would be better if these tsignore's can be resolved in the main code so they are no longer required.

--- a/src/lib/caching/stores/BanStore.ts
+++ b/src/lib/caching/stores/BanStore.ts
@@ -116,8 +116,6 @@ export class BanStore extends DataStore<Ban> {
 	 * @param data The data packet to add
 	 * @param cache If the data should be cached
 	 */
-	// eslint-disable-next-line @typescript-eslint/ban-ts-ignore
-	// @ts-ignore
 	protected _add(data: GuildBanAddDispatch['d']): Ban {
 		const existing = this.get(data.user.id);
 		// eslint-disable-next-line dot-notation

--- a/src/lib/caching/stores/GuildEmojiStore.ts
+++ b/src/lib/caching/stores/GuildEmojiStore.ts
@@ -87,8 +87,6 @@ export class GuildEmojiStore extends DataStore<GuildEmoji> {
 	 * Adds a new structure to this DataStore
 	 * @param data The data packet to add
 	 */
-	// eslint-disable-next-line @typescript-eslint/ban-ts-ignore
-	// @ts-ignore
 	protected _add(data: APIEmojiData): GuildEmoji {
 		const existing = this.get(data.id as string);
 		// eslint-disable-next-line dot-notation

--- a/src/lib/caching/stores/GuildInviteStore.ts
+++ b/src/lib/caching/stores/GuildInviteStore.ts
@@ -59,8 +59,6 @@ export class GuildInviteStore extends DataStore<Invite> {
 	 * @param data The data packet to add
 	 * @param cache If the data should be cached
 	 */
-	// eslint-disable-next-line @typescript-eslint/ban-ts-ignore
-	// @ts-ignore
 	protected _add(data: APIInviteData): Invite {
 		const existing = this.get(data.code);
 		// eslint-disable-next-line dot-notation

--- a/src/lib/caching/stores/GuildMemberStore.ts
+++ b/src/lib/caching/stores/GuildMemberStore.ts
@@ -118,8 +118,6 @@ export class GuildMemberStore extends DataStore<GuildMember> {
 	 * Adds a new structure to this DataStore
 	 * @param data The data packet to add
 	 */
-	// eslint-disable-next-line @typescript-eslint/ban-ts-ignore
-	// @ts-ignore
 	protected _add(data: MemberData): GuildMember {
 		const existing = this.get((data.user as APIUserData).id);
 		// eslint-disable-next-line dot-notation

--- a/src/lib/caching/stores/IntegrationStore.ts
+++ b/src/lib/caching/stores/IntegrationStore.ts
@@ -70,8 +70,6 @@ export class IntegrationStore extends DataStore<Integration> {
 	 * @param data The data packet to add
 	 * @param cache If the data should be cached
 	 */
-	// eslint-disable-next-line @typescript-eslint/ban-ts-ignore
-	// @ts-ignore
 	protected _add(data: APIIntegrationData): Integration {
 		const existing = this.get(data.id);
 		// eslint-disable-next-line dot-notation

--- a/src/lib/caching/stores/InviteStore.ts
+++ b/src/lib/caching/stores/InviteStore.ts
@@ -51,8 +51,6 @@ export class InviteStore extends DataStore<Invite> {
 	 * @param data The data packet to add
 	 * @param cache If the data should be cached
 	 */
-	// eslint-disable-next-line @typescript-eslint/ban-ts-ignore
-	// @ts-ignore
 	protected _add(data: APIInviteData): Invite {
 		const existing = this.get(data.code);
 		// eslint-disable-next-line dot-notation

--- a/src/lib/caching/stores/PresenceStore.ts
+++ b/src/lib/caching/stores/PresenceStore.ts
@@ -33,8 +33,6 @@ export class PresenceStore extends DataStore<Presence> {
 	 * Adds a new structure to this DataStore
 	 * @param data The data packet to add
 	 */
-	// eslint-disable-next-line @typescript-eslint/ban-ts-ignore
-	// @ts-ignore
 	protected _add(data: APIPresenceUpdateData): Presence {
 		const existing = this.get(data.user.id);
 		// eslint-disable-next-line dot-notation

--- a/src/lib/caching/stores/RoleStore.ts
+++ b/src/lib/caching/stores/RoleStore.ts
@@ -89,8 +89,6 @@ export class RoleStore extends DataStore<Role> {
 	 * Adds a new structure to this DataStore
 	 * @param data The data packet to add
 	 */
-	// eslint-disable-next-line @typescript-eslint/ban-ts-ignore
-	// @ts-ignore
 	protected _add(data: APIRoleData): Role {
 		const existing = this.get(data.id);
 		// eslint-disable-next-line dot-notation

--- a/src/lib/caching/stores/VoiceStateStore.ts
+++ b/src/lib/caching/stores/VoiceStateStore.ts
@@ -34,8 +34,6 @@ export class VoiceStateStore extends DataStore<VoiceState> {
 	 * Adds a new structure to this DataStore
 	 * @param data The data packet to add
 	 */
-	// eslint-disable-next-line @typescript-eslint/ban-ts-ignore
-	// @ts-ignore
 	protected _add(data: APIVoiceStatePartial): VoiceState {
 		const existing = this.get(data.user_id);
 		// eslint-disable-next-line dot-notation

--- a/src/lib/caching/stores/base/DataStore.ts
+++ b/src/lib/caching/stores/base/DataStore.ts
@@ -56,8 +56,8 @@ export class DataStore<S extends Structure> extends Cache<string, S> {
 	 * @param data The data packet to add
 	 * @param cache If the data should be cached
 	 */
-	protected _add(data: { id: string, [k: string]: any }): S {
-		const existing = this.get(data.id);
+	protected _add(data: Record<string, any>): S {
+		const existing = this.get(data.id as string);
 		// eslint-disable-next-line dot-notation
 		if (existing) return existing['_patch'](data);
 

--- a/src/lib/pieces/ActionStore.ts
+++ b/src/lib/pieces/ActionStore.ts
@@ -22,30 +22,30 @@ export class ActionStore extends Store<Action> {
 	 * @since 0.0.1
 	 */
 	public clear(): void {
-		for (const event of this.values()) this.delete(event);
+		for (const event of this.values()) this.remove(event);
 	}
 
 	/**
-	 * Deletes an action from the store.
+	 * Removes an action from the store.
 	 * @since 0.0.1
 	 * @param name An action object or a string representing the action name.
-	 * @returns Whether or not the delete was successful.
+	 * @returns Whether or not the removal was successful.
 	 */
-	public delete(name: Action | string): boolean {
+	public remove(name: Action | string): boolean {
 		const event = this.resolve(name);
 		if (!event) return false;
 		// eslint-disable-next-line dot-notation
 		event['_unlisten']();
-		return super.delete(event);
+		return super.remove(event);
 	}
 
 	/**
-	 * Sets up an action in our store.
+	 * Adds and sets up an action in our store.
 	 * @since 0.0.1
 	 * @param piece The event piece we are setting up
 	 */
-	public set(piece: Action): Action | null {
-		const event = super.set(piece);
+	public add(piece: Action): Action | null {
+		const event = super.add(piece);
 		if (!event) return null;
 		// eslint-disable-next-line dot-notation
 		event['_listen']();

--- a/src/lib/pieces/EventStore.ts
+++ b/src/lib/pieces/EventStore.ts
@@ -39,30 +39,30 @@ export class EventStore extends Store<Event> {
 	 * @since 0.0.1
 	 */
 	public clear(): void {
-		for (const event of this.values()) this.delete(event);
+		for (const event of this.values()) this.remove(event);
 	}
 
 	/**
-	 * Deletes an event from the store.
+	 * Removes an event from the store.
 	 * @since 0.0.1
 	 * @param name An event object or a string representing the event name.
-	 * @returns Whether or not the delete was successful.
+	 * @returns Whether or not the removal was successful.
 	 */
-	public delete(name: Event | string): boolean {
+	public remove(name: Event | string): boolean {
 		const event = this.resolve(name);
 		if (!event) return false;
 		// eslint-disable-next-line dot-notation
 		event['_unlisten']();
-		return super.delete(event);
+		return super.remove(event);
 	}
 
 	/**
-	 * Sets up an event in our store.
+	 * Adds and sets up an event in our store.
 	 * @since 0.0.1
 	 * @param piece The event piece we are setting up
 	 */
-	public set(piece: Event): Event | null {
-		const event = super.set(piece);
+	public add(piece: Event): Event | null {
+		const event = super.add(piece);
 		if (!event) return null;
 		// eslint-disable-next-line dot-notation
 		event['_listen']();

--- a/src/lib/pieces/base/AliasStore.ts
+++ b/src/lib/pieces/base/AliasStore.ts
@@ -34,28 +34,28 @@ export class AliasStore<V extends AliasPiece> extends Store<V> {
 	}
 
 	/**
-	 * Sets up an AliasPiece in our store.
+	 * Adds and sets up an AliasPiece in our store.
 	 * @since 0.0.1
 	 * @param piece The piece we are setting up
 	 */
-	public set(piece: V): V | null {
-		const aliasPiece = super.set(piece);
+	public add(piece: V): V | null {
+		const aliasPiece = super.add(piece);
 		if (!aliasPiece) return null;
 		for (const alias of aliasPiece.aliases) this.aliases.set(alias, aliasPiece);
 		return aliasPiece;
 	}
 
 	/**
-	 * Deletes an AliasPiece from the store.
+	 * Removes an AliasPiece from the store.
 	 * @since 0.0.1
 	 * @param name An AliasPiece object or a string representing an AliasPiece or alias name
-	 * @returns Whether or not the delete was successful.
+	 * @returns Whether or not the removal was successful.
 	 */
-	public delete(name: V | string): boolean {
+	public remove(name: V | string): boolean {
 		const aliasPiece = this.resolve(name) as V | null;
 		if (!aliasPiece) return false;
 		for (const alias of aliasPiece.aliases) this.aliases.delete(alias);
-		return super.delete(aliasPiece);
+		return super.remove(aliasPiece);
 	}
 
 	/**

--- a/src/lib/pieces/base/Piece.ts
+++ b/src/lib/pieces/base/Piece.ts
@@ -96,7 +96,7 @@ export class Piece {
 	 */
 	public unload(): boolean {
 		this.client.emit('pieceUnloaded', this);
-		return this.store.delete(this);
+		return this.store.remove(this);
 	}
 
 	/**

--- a/src/lib/pieces/base/Store.ts
+++ b/src/lib/pieces/base/Store.ts
@@ -91,7 +91,7 @@ export class Store<V extends Piece> extends Cache<string, V> {
 			const loaded = await import(loc) as { default: PieceConstructor<V> } | PieceConstructor<V>;
 			const LoadedPiece = 'default' in loaded ? loaded.default : loaded;
 			if (!isClass(LoadedPiece)) throw new TypeError('The exported structure is not a class.');
-			piece = this.set(new LoadedPiece(this, directory, file));
+			piece = this.add(new LoadedPiece(this, directory, file));
 		} catch (error) {
 			this.client.emit('wtf', `Failed to load file '${loc}'. Error:\n${error.stack || error}`);
 		}
@@ -115,20 +115,18 @@ export class Store<V extends Piece> extends Cache<string, V> {
 	}
 
 	/**
-	 * Sets up a piece in our store.
+	 * Adds and sets up a piece in our store.
 	 * @since 0.0.1
 	 * @param piece The piece we are setting up
 	 */
-	// eslint-disable-next-line @typescript-eslint/ban-ts-ignore
-	// @ts-ignore
-	public set(piece: V): V | null {
+	public add(piece: V): V | null {
 		if (!(piece instanceof this.holds)) {
 			this.client.emit('error', `Only ${this} may be stored in this Store.`);
 			return null;
 		}
 
 		// Remove any previous piece named the same
-		this.delete(piece.name);
+		this.remove(piece.name);
 
 		// Emit pieceLoaded event, set to the cache, and return it
 		this.client.emit('pieceLoaded', piece);
@@ -137,16 +135,32 @@ export class Store<V extends Piece> extends Cache<string, V> {
 	}
 
 	/**
-	 * Deletes a piece from the store.
+	 * Removes a piece from the store.
 	 * @since 0.0.1
 	 * @param name A piece instance or a string representing a piece or alias name
-	 * @returns Whether or not the delete was successful.
+	 * @returns Whether or not the removal was successful.
 	 */
-	public delete(name: V | string): boolean {
+	public remove(name: V | string): boolean {
 		const piece = this.resolve(name);
 		if (!piece) return false;
 		super.delete(piece.name);
 		return true;
+	}
+
+	/**
+	 * The overriden set method, this will always throw.
+	 * @internal
+	 */
+	public set(): never {
+		throw new Error('Cannot set in this Store.');
+	}
+
+	/**
+	 * The overriden delete method, this will always throw.
+	 * @internal
+	 */
+	public delete(): never {
+		throw new Error('Cannot delete in this Store.');
 	}
 
 	/**

